### PR TITLE
[FE] Automate check that MCDA is not available for guest/low-rights user

### PR DESCRIPTION
**Pull request by amurKontur**
- created all buttons check in toolbar for user with no rights and guest
- created population popup check with 4 cases: small, normal, huge population and popup near equator
- created logics for sharing auth state of files
- created auth version of all current tests that broaden coverage
- fixed other tests issues with async behaviour and etc

### CODE REVIEW

File: e2e/fixtures/test-options.ts [hunk 1]
```diff
@@ -1,13 +1,42 @@
-import { test as basis } from '@playwright/test';
+import { test as basis, expect } from '@playwright/test';
 import { PageManager } from '../page-objects/pageManager';
+import type { Project } from '../page-objects/helperBase.ts';
 
 export type TestOptions = {
   pageManager: PageManager;
+  project: Project;
+  loginWithNoRightsUser: string;
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply e2e/fixtures/test-options.ts]: this code is out-of-date

File: e2e/page-objects/mapPage.ts [hunk 2]
```diff
@@ -1,7 +1,36 @@
-import { HelperBase, Project } from './helperBase';
 import { expect } from '@playwright/test';
+import { HelperBase } from './helperBase';
+import type { Project } from './helperBase';
 
 export class MapCanvas extends HelperBase {
+  /**
+   * This method devides text by ':' and expects 2 parts. 1 part - label with expectedLabel text and 2 - number. Number is checked for being not NaN. If value can not be 0, then it will be checked for being not 0 too.
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply e2e/page-objects/mapPage.ts]: fixed

File: e2e/page-objects/mapPage.ts [hunk 3]
```diff
@@ -22,4 +51,69 @@
...
+    latitude: number,
+    longitude: number,
+    project: Project,
+  ) {
+    const urlForPlace =
+      project.env === 'prod'
+        ? `${project.url}/active/?map=${zoom}/${latitude}/${longitude}`
+        : `${project.url}&map=${zoom}/${latitude}/${longitude}`;
+    await this.page.goto(urlForPlace, { waitUntil: 'domcontentloaded' });
+    await this.waitForZoom();
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply e2e/page-objects/mapPage.ts:69]: the suggested variant is incorrect, it does not include information about project url so it won't work. i'd better leave as it is

File: e2e/page-objects/helperBase.ts [hunk 4]
```diff
@@ -11,6 +11,7 @@
   title: string;
   hasCookieBanner: boolean;
   hasAtlasBanner: boolean;
+  authUrl: string;
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply e2e/page-objects/helperBase.ts:18]: it is autotest itself, you are wrong

File: e2e/page-objects/helperBase.ts [hunk 5]
```diff
@@ -19,25 +20,46 @@
...
     this.page = page;
   }
+  /**
+   * This method opens up a project like disaster-ninja, atlas, etc. After that it checks the title to correspond to this project and accepts cookies if needed
+   * @param project object with details about project to open like name, url, title, etc.
+   * @param hasNoCookiesAccepted has browser got cookies already accepted for this project
+   */
 
-  async openProject(project: Project) {
+  async openProject(project: Project, hasNoCookiesAccepted: boolean = true) {
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply e2e/page-objects/helperBase.ts]: it is autotest itself, you are wrong

File: e2e/page-objects/helperBase.ts [hunk 6]
```diff
@@ -19,25 +20,46 @@
...
 
   async closeAtlasBanner(project: Project) {
     if (project.hasAtlasBanner) {
       await this.page.waitForSelector('[title="Intercom live chat banner"]');
-      const frame = this.page.frameLocator('[title="Intercom live chat banner"]');
-      await frame.getByLabel('Close').locator('> :first-child').click();
+      await this.page
+        .frameLocator('[title="Intercom live chat banner"]')
+        .getByLabel('Close')
+        .click();
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply e2e/page-objects/helperBase.ts]: it is autotest itself, you are wrong

File: e2e/page-objects/helperBase.ts [hunk 7]
```diff
@@ -19,25 +20,46 @@
...
+   */
+  async getTextsFromAllLocators(locators: Locator[]) {
+    const textsArray: string[] = [];
+    for (const locator of locators) {
+      const text = await locator.textContent();
+      expect(text).not.toBeNull();
+      textsArray.push(text!);
+    }
+    return textsArray;
+  }
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply e2e/page-objects/helperBase.ts]: it is autotest itself, you are wrong

File: playwright.config.ts [hunk 8]
```diff
@@ -13,10 +13,10 @@
  * See https://playwright.dev/docs/test-configuration.
  */
 export default defineConfig({
-  globalTimeout: process.env.CI ? 300000 : 240000,
+  globalTimeout: process.env.CI ? 900000 : 600000,
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply playwright.config.ts:16]: it is autotest itself, you are wrong

File: playwright.config.ts [hunk 9]
```diff
@@ -13,10 +13,10 @@
...
   timeout: process.env.CI ? 120000 : 60000,
   expect: {
-    timeout: process.env.CI ? 10000 : 5000,
+    timeout: process.env.CI ? 10000 : 7000,
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply playwright.config.ts:19]: it is autotest itself, you are wrong

File: playwright.config.ts [hunk 10]
```diff
@@ -42,17 +42,36 @@
 
   /* Configure projects for major browsers */
   projects: [
+    {
+      name: 'setup',
+      testMatch: 'auth.setup.ts',
+    },
+
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply playwright.config.ts:49]: it is autotest itself, you are wrong

File: playwright.config.ts [hunk 11]
```diff
@@ -42,17 +42,36 @@
...
     {
       name: 'chromium',
       use: { ...devices['Desktop Chrome'] },
+      testIgnore: /WithUser/,
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply playwright.config.ts:53]: it is autotest itself, you are wrong

File: playwright.config.ts [hunk 12]
```diff
@@ -42,17 +42,36 @@
...
       use: { ...devices['Desktop Safari'] },
-      testIgnore: 'location.spec.ts',
+      testIgnore: [/WithUser/, 'location.spec.ts'],
+    },
+
+    {
+      name: 'chromium_logged_in',
+      use: { ...devices['Desktop Chrome'], storageState: 'e2e/.auth/user.json' },
+      dependencies: ['setup'],
+      testMatch: /WithUser/,
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply playwright.config.ts:66]: it is autotest itself, you are wrong

File: playwright.config.ts [hunk 13]
```diff
@@ -42,17 +42,36 @@
...
     },
 
+    {
+      name: 'webkit_logged_in',
+      use: { ...devices['Desktop Safari'], storageState: 'e2e/.auth/user.json' },
+      dependencies: ['setup'],
+      testMatch: /WithUser/,
+    },
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply playwright.config.ts:74]: it is autotest itself, you are wrong

File: e2e/page-objects/mapPage.ts [hunk 14]
```diff
@@ -1,7 +1,36 @@
-import { HelperBase, Project } from './helperBase';
 import { expect } from '@playwright/test';
+import { HelperBase } from './helperBase';
+import type { Project } from './helperBase';
 
 export class MapCanvas extends HelperBase {
+  /**
+   * This method divides text by ':' and expects 2 parts. 1 part - label with expectedLabel text and 2 - number. Number is checked for being not NaN. If value can not be 0, then it will be checked for being not 0 too.
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply e2e/page-objects/mapPage.ts]: I don't think it is a typo, more like text enhancement.

File: e2e/page-objects/mapPage.ts [hunk 15]
```diff
@@ -22,4 +51,69 @@
...
+  async goToSpecificAreaByUrl(
+    zoom: number,
+    latitude: number,
+    longitude: number,
+    project: Project,
+  ) {
+    const urlForPlace =
+      project.env === 'prod'
+        ? `${project.url}/active/?map=${zoom}/${latitude}/${longitude}`
+        : `${project.url}&map=${zoom}/${latitude}/${longitude}`;
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply e2e/page-objects/mapPage.ts:67]: duplicate

File: e2e/fixtures/test-options.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (amurKontur) [reply e2e/fixtures/test-options.ts]: can you fix it yourself, just delete loginWithNoRightsUser: string; line as it is not about to be used

File: e2e/page-objects/mapPage.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (amurKontur) [reply e2e/page-objects/mapPage.ts]: you can fix it yourself if you have possibility

File: e2e/populationPopup.spec.ts [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+        pageManager: pageManager,
+        project: project,
+        zoom: 15.307,
+        latitude: -0.0191,
+        longitude: 10.3915,
+        x: 200,
+        y: 200,
+      });
+    });
+  });
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply e2e/populationPopup.spec.ts:91]: callback hell. bad idea

File: e2e/populationPopup.spec.ts [hunk 17]
```diff
@@ -0,0 +1,92 @@
...
+  await inputData.pageManager.atMap.goToSpecificAreaByUrl(
+    inputData.zoom,
+    inputData.latitude,
+    inputData.longitude,
+    inputData.project,
+  );
+  await inputData.pageManager.atMap.clickPlaceOnMapView(inputData.x, inputData.y);
+  await inputData.pageManager.atMap.checkPopulationPopupData();
+  await inputData.pageManager.atMap.closePopulationPopup();
+};
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply e2e/populationPopup.spec.ts:33]: it is autotest, not app

File: e2e/page-objects/mapPage.ts [hunk 18]
```diff
@@ -22,4 +47,69 @@ export class MapCanvas extends HelperBase {
...
+  async closePopulationPopup() {
+    await this.page
+      .locator('.maplibregl-popup-content')
+      .getByLabel('Close popup')
+      .click();
+    await expect(
+      this.page.locator('.maplibregl-popup-content', { hasText: 'Population' }),
+    ).not.toBeVisible();
+    await expect(this.page.locator('#map-view')).toBeVisible();
+  }
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply e2e/page-objects/mapPage.ts:114]: it is locator, it should be so for this method

File: e2e/page-objects/mapPage.ts [hunk 19]
```diff
@@ -1,7 +1,32 @@
...
+   */
+
+  checkPopupData(resultedText: string, expectedLabel: string, valueCanNotBe0: boolean) {
+    const arr = resultedText.split(':');
+    expect(arr.length).toEqual(2);
+    expect(arr[0].trim()).toEqual(expectedLabel);
+    const value = Number(arr[1].trim());
+    expect(value).not.toBeNaN();
+    if (valueCanNotBe0) expect(value).not.toEqual(0.0);
+  }
```
[the remark this answers is not in the document]
Comment (amurKontur) [reply e2e/page-objects/mapPage.ts]: fixed

**Review by dqunbp — APPROVED** [review]

File: e2e/fixtures/test-options.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (amurKontur) [reply e2e/fixtures/test-options.ts]: fixed

File: e2e/page-objects/mapPage.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (amurKontur) [reply e2e/page-objects/mapPage.ts]: fixed in another comment

File: e2e/populationPopup.spec.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (amurKontur) [reply e2e/populationPopup.spec.ts:91]: i don't think it is needed here, separate test is better

**Pull request merged**